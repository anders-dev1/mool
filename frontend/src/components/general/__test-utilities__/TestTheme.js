export const testTheme = {
    textColor: "black",
    primaryColor: "#007BFF",
    primaryHover: "rgba(0,123,255,0.33)",
    successColor: "#28A745",
    errorColor: "#F02849",
    backgroundColor: "#F0F2F5",
    buttonTextColor: "#FFFFFF",
    dividerColor: "#C7CBD1",
    inputColor: "#99A5B0",
    spinnerColor: "#FFFFFF transparent transparent transparent",
    font: "Acumin Pro, Arial, sans-serif !important;"
};