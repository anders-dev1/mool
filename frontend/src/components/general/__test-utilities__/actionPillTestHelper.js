import {screen} from "@testing-library/react";

export function findActionPillByIcon(icon) {
    return screen.getByText(icon).closest('div');
}

export function findActionPillNumberByIcon(icon) {
    const pill = findActionPillByIcon(icon);
    const numberP = pill.lastChild.lastChild;
    return numberP.textContent;
}