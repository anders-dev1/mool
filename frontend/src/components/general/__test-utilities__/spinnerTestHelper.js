export function findSpinner(container){
    return container.getElementsByClassName('lds-ring')[0];
}