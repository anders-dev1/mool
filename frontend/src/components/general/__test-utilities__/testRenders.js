import React from 'react'
import {render} from '@testing-library/react'
import {Provider} from 'react-redux'
import {setupStore, store} from "../../../redux/store";
import {testTheme} from "./TestTheme";
import {ThemeProvider} from "styled-components";

// This is used to set up the store for tests.
export function renderWithProviders(
    ui,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        testStore = setupStore(preloadedState),
        ...renderOptions
    } = {}
) {
    function Wrapper({children}) {
        return(
            <Provider store={testStore}>
                <ThemeProvider theme={testTheme}>
                {children}
                </ThemeProvider>
            </Provider>
        );
    }

    return {testStore, ...render(ui, {wrapper: Wrapper, ...renderOptions})}
}

// This is used with the react-test-renderer to render dom in tests.
export function internalsForRenderer(
    ui,
    {
        preloadedState = {},
        testStore = setupStore(preloadedState)
    } = {}
){
    function Wrapper(children) {
        return(
            <Provider store={testStore}>
                <ThemeProvider theme={testTheme}>
                    {children}
                </ThemeProvider>
            </Provider>
        );
    }

    return Wrapper(ui);
}