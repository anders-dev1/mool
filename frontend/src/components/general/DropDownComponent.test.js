import DropDown from "./DropDownComponent";
import {ThemeProvider} from "styled-components";
import {testTheme} from "./__test-utilities__/TestTheme";
import {fireEvent, render, screen} from "@testing-library/react";

const testOptions = [
    {label: "First", value: "first value"},
    {label: "Second", value: "second value"},
    {label: "Third", value: "third value"}
];

test('when in initial state, renders correctly', () => {
   const {container} = render(
       <ThemeProvider theme={testTheme}>
         <DropDown options={testOptions} initiallySelectedIndex={0} placeholder={"test placeholder"}/>
       </ThemeProvider>
   );

   expect(container).toMatchSnapshot('Default state');
});

test('when dropped down, renders correctly', () => {
   // Arrange
   const {container} = render(
       <ThemeProvider theme={testTheme}>
          <DropDown options={testOptions} initiallySelectedIndex={0}/>
       </ThemeProvider>
   );

   const dropDown = container.querySelector('div').firstChild;

   // Act
   fireEvent.click(dropDown);

   // Assert
   expect(container).toMatchSnapshot('Dropped down');
});

test('when initiallySelectedIndex provided, will set selectedIndex to what was provided', () => {
    // Arrange
    const index = 2;
    render(
        <ThemeProvider theme={testTheme}>
            <DropDown options={testOptions} initiallySelectedIndex={index}/>
        </ThemeProvider>
    );

    // Assert
    const option = screen.getByText(testOptions[index].label);
    expect(option).toBeDefined();
});

describe('option clicked', () =>{
    test('when option is not the one currently selected, should change selected index', () => {
        // Arrange
        const initiallySelectedIndex = 0;
        const indexToSelect = 1;
        const {container} = render(
            <ThemeProvider theme={testTheme}>
                <DropDown options={testOptions} initiallySelectedIndex={initiallySelectedIndex}/>
            </ThemeProvider>
        );

        // Act
        const dropDown = container.querySelector('div').firstChild;
        fireEvent.click(dropDown);
        const optionToSelect = screen.getByText(testOptions[indexToSelect].label);
        fireEvent.click(optionToSelect);

        // Assert
        // At this point the dropdown should have collapsed and only the newly selected option should be visible.
        const option = screen.getByText(testOptions[indexToSelect].label);
        expect(option).toBeDefined();
    });

    test('when option is not the one currently selected and changeEvent is provided, should call the changeEvent with option', () => {
        // Arrange
        const mockCallBack = jest.fn();
        const initiallySelectedIndex = 0;
        const indexToSelect = 1;
        const {container} = render(
            <ThemeProvider theme={testTheme}>
                <DropDown options={testOptions} initiallySelectedIndex={initiallySelectedIndex} changedEvent={mockCallBack}/>
            </ThemeProvider>
        );

        // Act
        const dropDown = container.querySelector('div').firstChild;
        fireEvent.click(dropDown);
        const optionToSelect = screen.getByText(testOptions[indexToSelect].label);
        fireEvent.click(optionToSelect);

        // Assert
        expect(mockCallBack).toHaveBeenCalledWith(testOptions[indexToSelect]);
    });

    test('when option is the same as the current index, should not call changeEvent', () => {
        // Arrange
        const mockCallBack = jest.fn();
        const initiallySelectedIndex = 0;
        const {container} = render(
            <ThemeProvider theme={testTheme}>
                <DropDown options={testOptions} initiallySelectedIndex={initiallySelectedIndex} changedEvent={mockCallBack}/>
            </ThemeProvider>
        );

        // Act
        const dropDown = container.querySelector('div').firstChild;
        fireEvent.click(dropDown);
        const optionToSelect = screen.getAllByText(testOptions[initiallySelectedIndex].label)[1];
        fireEvent.click(optionToSelect);

        // Assert
        expect(mockCallBack).not.toHaveBeenCalled();
    });
});