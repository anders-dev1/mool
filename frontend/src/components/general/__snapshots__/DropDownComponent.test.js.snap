// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when in initial state, renders correctly: Default state 1`] = `
<div>
  <div
    class="sc-bczRLJ cseizD"
  >
    <div
      class="sc-gsnTZi tvYfs"
    >
      <div
        class="sc-dkzDqf cSTGXO"
      >
        <p>
          First
        </p>
        <img
          alt="expand comments"
          src="expand.svg"
          style="transform: translateY(-2px);"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`when dropped down, renders correctly: Dropped down 1`] = `
<div>
  <div
    class="sc-bczRLJ cseizD"
  >
    <div
      class="sc-gsnTZi ionnPL"
    >
      <div
        class="sc-dkzDqf cSTGXO"
      >
        <p>
          First
        </p>
        <img
          alt="expand comments"
          src="expand.svg"
          style="transform: translateY(-2px);"
        />
      </div>
      <div
        class="sc-hKMtZM eKGUlp"
      >
        <div
          class="sc-eCYdqJ gfgMTH"
        >
          <p>
            First
          </p>
        </div>
        <div
          class="sc-eCYdqJ cliWDj"
        >
          <p>
            Second
          </p>
        </div>
        <div
          class="sc-eCYdqJ cliWDj"
        >
          <p>
            Third
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;